src/applications/appeals/10182/containers/FormApp.jsx

```diff
@@ -105,19 +108,40 @@ export const FormApp = ({
 
   useEffect(
     () => {
-      // storage flag to dynamically add part3 migation for redirect to request
-      // extension page
-      window.sessionStorage.setItem(SHOW_PART3_STORAGE, showPart3);
-
+      // Checking returnUrl to ensure the redirect occurs _after_ the save-in-
+      // progress response has loaded; and saving this check to form data to
+      // ensure that this check & redirect only occurs once & the state is saved
+      // in the form data so it doesn't occur again if the Veteran returns later
+      if (showPart3 && returnUrl && !formData[SHOW_PART3_REDIRECT]) {
+        const needsRedirect = checkRedirect(returnUrl);
+        // Using "redirected" for the resulting page to show an info alert so
+        // the Veteran knows why they were redirected.
+        setFormData({
+          ...formData,
+          [SHOW_PART3_REDIRECT]: needsRedirect ? 'redirected' : 'not-needed',
+        });
+        // Redirect back to part 3 question if Veteran is on or past the
+        // contestable issues page
+        if (needsRedirect) {
+          // Using a setTimeout because the save-in-progress redirect pushes
+          // the returnUrl page to the router _after_ this event.
+          setTimeout(() => {
+            router.push('/extension-request');
+          }, 100);
```

Comment:
something i was noodling on — mostly because i don't have a full understanding of sequence of events/actions yet — i was wondering what happens if we update the `returnUrl` to this value instead of pushing directly with the router? would the app do anything / would it detect the change and redirect? or would it only redirect with the old value because by the time this code runs, it's too late to update the `returnUrl`? 🤔